packages/block-editor/src/components/block-list/index.native.js

```diff
@@ -311,42 +312,55 @@ export class BlockList extends Component {
 }
 
 export default compose( [
-	withSelect( ( select, { rootClientId, __experimentalMoverDirection } ) => {
-		const {
-			getBlockCount,
-			getBlockOrder,
-			getSelectedBlockClientId,
-			isBlockInsertionPointVisible,
-			getSettings,
-			getBlockHierarchyRootClientId,
-		} = select( 'core/block-editor' );
-
-		const isStackedHorizontally =
-			__experimentalMoverDirection === 'horizontal';
-
-		const selectedBlockClientId = getSelectedBlockClientId();
-		const blockClientIds = getBlockOrder( rootClientId );
-
-		const isReadOnly = getSettings().readOnly;
-
-		const rootBlockId = getBlockHierarchyRootClientId(
-			selectedBlockClientId
-		);
-		const hasRootInnerBlocks = !! getBlockCount( rootBlockId );
-
-		const isFloatingToolbarVisible =
-			!! selectedBlockClientId && hasRootInnerBlocks;
-
-		return {
-			blockClientIds,
-			blockCount: getBlockCount( rootClientId ),
-			isBlockInsertionPointVisible: isBlockInsertionPointVisible(),
-			isReadOnly,
-			isRootList: rootClientId === undefined,
-			isFloatingToolbarVisible,
-			isStackedHorizontally,
-		};
-	} ),
+	withSelect(
+		(
+			select,
+			{ rootClientId, __experimentalMoverDirection, filterInnerBlocks }
+		) => {
+			const {
+				getBlockCount,
+				getBlockOrder,
+				getSelectedBlockClientId,
+				isBlockInsertionPointVisible,
+				getSettings,
+				getBlockHierarchyRootClientId,
+			} = select( 'core/block-editor' );
+
+			const isStackedHorizontally =
+				__experimentalMoverDirection === 'horizontal';
+
+			const selectedBlockClientId = getSelectedBlockClientId();
+
+			let blockClientIds = getBlockOrder( rootClientId );
+
+			// Display only block which fulfill the condition in passed `filterInnerBlocks` function
+			if ( filterInnerBlocks ) {
+				blockClientIds = filterInnerBlocks(
+					getBlockOrder( rootClientId )
+				);
+			}
```

Comment:
Hmm i believe we could refactor it a bit since we don't need to call `gerBlockOrder` twice if the `filterInnerBlocks` is true. Maybe we could do something like:
```js
const blockClientIds = filterInnerBlocks ? filterInnerBlocks(getBlockOrder( rootClientId )) : getBlockOrder( rootClientId );
```